scripts: dts: support multiple binding dirs in new scripts

mike-scott opened the pull request:
gen_defines.py and edtlib.py were recently added in
commit 62d57414762d ("dts: Add new DTS/binding parser").

The old extract_dts_includes.py script allowed for multiple
dts bindings dirs.  Let's add that functionality to the new
scripts as well.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
scripts: dts: support multiple binding dirs in new scripts

gen_defines.py and edtlib.py were recently added in
commit 62d57414762d ("dts: Add new DTS/binding parser").

The old extract_dts_includes.py script allowed for multiple
dts bindings dirs.  Let's add that functionality to the new
scripts as well.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@ulfalizer NOTE: the change to the new DTS parsing scripts breaks users of the old extract_dts_includes.py using multiple bindings directories.

mike-scott commented:
Not sure if this should be marked a "bug" or not.  Maybe a regression?

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -101,8 +101,8 @@ def __init__(self, dts, bindings_dir):
           Path to device tree .dts file
 
         bindings_dir:
```
Could call this this `binding_dirs` maybe (`bindings_dirs` is a bit of a mouthfull).

ulfalizer replied:
Heh... looks like someone else thought `bindings_dirs` was a bit much. `binding_dirs` might be a happy medium.

pabigot replied:
Not if the directory name is `bindings`.

ulfalizer replied:
I was thinking "binding directories" as in "directories with bindings".

It's pretty meh though. Just put the `s`s where you feel like it. :P

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -101,8 +101,8 @@ def __init__(self, dts, bindings_dir):
         bindings_dir:
-          Path to directory containing bindings, in YAML format. This directory
-          is recursively searched for .yaml files.
+          Paths to directory containing bindings, in YAML format. These
```
`List of paths to directories containing...` maybe.

galak commented:
> Not sure if this should be marked a "bug" or not. Maybe a regression?

marked it 'bug' as well, since it is one. :)

galak commented:
LGTM, will let @ulfalizer have some time to review.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1013,10 +1013,15 @@ def _binding_paths(bindings_dir):
-            for root, _, filenames in os.walk(bindings_dir)
-                for filename in filenames
-                    if filename.endswith(".yaml")]
+    binding_files = []
```
Could call it `binding_paths` maybe. Won't overlap with the function due to the `_`.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -101,8 +101,8 @@ def __init__(self, dts, bindings_dir):
           Path to device tree .dts file
 
         bindings_dir:
```
might want to rename this to `bindings_dirs`

mike-scott replied:
I actually had this changed all through-out (in both files).  But I thought it made the changes look worse than they actually are.

mike-scott replied:
IE: since it was a regression I wanted to keep it simple

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -40,7 +40,7 @@ def main():
     out_comment("Generated by gen_defines.py", blank_before=False)
     out_comment("DTS input file: " + args.dts, blank_before=False)
-    out_comment("Directory with bindings: " + args.bindings_dir,
+    out_comment("Directories with bindings: '{}'".format(args.bindings_dir),
```
Could do `", ".join(args.binding_dirs)` maybe, to make it consistent with how lists are shown in some other places.

ulfalizer replied:
Thinking

```python
out_comment("Directories with bindings: " + ", ".join(args.binding_dirs))
```

ulfalizer reviewed (changes requested):
Fine with me, even though I don't know where we use it. Just some nits.

Try running `./testedtlib.py` in `scripts/dts` too, just to be sure. Not sure if this hits anything it covers though. It'll be run automatically in CI later.

pabigot commented on scripts/dts/gen_defines.py:
```diff
@@ -114,8 +114,9 @@ def parse_args():
-                        help="directory with bindings in YAML format")
+    parser.add_argument("--bindings-dir", nargs='+', required=True,
+                        help="directory with bindings in YAML format, "
+                        "we allow multiple")
```
I think `--bindings-dirs`.  Four of the five scripts that use `nargs='+'` use plural nouns in the name.

ulfalizer replied:
`--binding-dirs` fine too (just moving the `s`)? The double `s` keeps tripping me up.

pabigot replied:
Not really fine with me, no.  The directory is named `bindings`.  When there was one of them, `--bindings-dir` was right.  When there are multiple it should be `--bindings-dirs`.  Name consistency is important for preservation of sanity.

pabigot replied:
Though this is a bikeshed argument.  I'll live with either (but will curse if my file isn't found because I put it in a directory named `binding` that cmake didn't look at).

ulfalizer replied:
The script doesn't care what the directory is called at least. `--bindings-dir` just meant "this is a directory with bindings" to me, not "this is the directory with the naming bindings/".

mike-scott commented:
> Fine with me, even though I don't know where we use it. Just some nits.
>
>Try running ./testedtlib.py in scripts/dts too, just to be sure. Not sure if this hits anything it covers though. It'll be run automatically in CI later.

Ah nice test!  This looks broken somehow.. let me dig into it.

mike-scott commented:
> Fine with me, even though I don't know where we use it. Just some nits.
>
>Try running ./testedtlib.py in scripts/dts too, just to be sure. Not sure if this hits anything it covers though. It'll be run automatically in CI later.

@ulfalizer Ah nice test!  This commit breaks it somehow.  Let me dig into it.

pabigot reviewed (changes requested):
Yes, please.  I haven't needed this yet, but will.

ulfalizer commented:
> > Fine with me, even though I don't know where we use it. Just some nits.
> > Try running ./testedtlib.py in scripts/dts too, just to be sure. Not sure if this hits anything it covers though. It'll be run automatically in CI later.
> 
> Ah nice test! This looks broken somehow.. let me dig into it.

It's a bit stone age and mostly just compares `repr()` output, so might be the change there. It's a decent way to cover a lot of code with little test code at least.

pabigot commented:
> Fine with me, even though I don't know where we use it. Just some nits.

Per `dts.cmake` it should take effect if any board or application `dts` directory has a `bindings` subdirectory.  For Particle devices this could have been used to describe the pin specifications for the board-specific antenna switch.

ulfalizer commented:
> > Fine with me, even though I don't know where we use it. Just some nits.
> 
> Per `dts.cmake` it should take effect if any board or application `dts` directory has a `bindings` subdirectory. For Particle devices this could have been used to describe the pin specifications for the board-specific antenna switch.

Alright, thanks for looking into it.

Currently there's an error if two different binding files have the same compatible (was a warning with the old scripts). Wonder if it would break it.

ulfalizer commented:
@mike-scott 
Just thought of one thing it breaks. Will need to do

```
diff --git a/scripts/dts/testedtlib.py b/scripts/dts/testedtlib.py
index 094e5771ae..150367284a 100755
--- a/scripts/dts/testedtlib.py
+++ b/scripts/dts/testedtlib.py
@@ -30,7 +30,7 @@ def run():
             fail("not equal (expected value last):\n'{}'\n'{}'"
                  .format(actual, expected))
 
-    edt = edtlib.EDT("test.dts", "test-bindings")
+    edt = edtlib.EDT("test.dts", ["test-bindings"])
 
     #
     # Test interrupts
```

mike-scott commented:
@ulfalizer that was it.  With that change the test passes.  I'll add that to this PR.

Also, should we have a test for 2 bindings dirs (gah dual plurals) ?  IE: I can make a `test-bindings2` dir and move `interrupt-3-cell.yaml` there as well.

ulfalizer commented:
> Also, should we have a test for 2 bindings dirs (gah dual plurals) ? IE: I can make a `test-bindings2` dir and move `interrupt-3-cell.yaml` there as well.

Sounds good to me. Would be nice to have it covered.

mike-scott commented:
updated to cover comments.. and yes, I went with `bindings_dirs`.

mike-scott commented:
I notice checkpatch barfs all over the test DTS files..  I basically copy/pasted them with the binding name changed.

ulfalizer reviewed (changes requested):
I was thinking just moving one of the bindings over to `test-bindings-2/`, e.g. `test-bindings-2/interrupt-3-cell.yaml`, and modifying the existing tests to pass. Might be overkill to copy a bunch of tests for it.

(Could have a dedicated `test-bindings-2/find-me.yaml` binding and a new test as well, but I could fix that up later.)

That means we only test two bindings dirs and not a single bindings dirs, but it's probably alright. It'll be obvious if a single dir breaks.

ulfalizer commented:
> I notice checkpatch barfs all over the test DTS files.. I basically copy/pasted them with the binding name changed. (all warnings btw)

Yeah, it's trying to match the `compatible = "..."` strings in `test.dts` against the bindings in `dts/bindings/` (instead of the bindings in `test-bindings/`). Noticed it earlier, but didn't find a way to fix it.

mike-scott commented:
> I was thinking just moving one of the bindings over to `test-bindings-2/`, e.g. `test-bindings-2/interrupt-3-cell.yaml`, and modifying the existing tests to pass. Might be overkill to copy a bunch of tests for it.
> 
> (Could have a dedicated `test-bindings-2/find-me.yaml` binding and a new test as well, but I could fix that up later.)
> 
> That means we only test two bindings dirs and not a single bindings dirs, but it's probably alright. It'll be obvious if a single dir breaks.

@ulfalizer slightly confused.  I did make a test-binding2 dir and made a copy of interrupt-3-cell.yaml as interrupt-3-cell-2dir.yaml (with compatible name change: interrupt-3-cell-2dir)

I should probably cut down the size of the test-2dir.dts file so that it only has the needed data for the tests added.

But did you need something else changed there?

ulfalizer commented:
> @ulfalizer slightly confused. I did make a test-binding2 dir and made a copy of interrupt-3-cell.yaml as interrupt-3-cell-2dir.yaml (with compatible name change: interrupt-3-cell-2dir)
> 
> I should probably cut down the size of the test-2dir.dts file so that it only has the needed data for the tests added.
> 
> But did you need something else changed there?

I mean only moving `interrupt-3-cell.yaml` over to `test-bindings-2/interrupt-3-cell.yaml` (the filename and contents could be the same), and then changing

```python
edt = edtlib.EDT("test.dts", ["test-bindings"])
```

to

```python
edt = edtlib.EDT("test.dts", ["test-bindings", "test-bindings-2"])
```

No need to add any new files, or add/remove any lines in `testedtlib.py`. After that, there's coverage for two binding directories.

ulfalizer commented:
And yeah, that's piggybacking on an existing test. Could have a dedicated `test-bindings2/find-me-yaml` that matches with a new `compatible = "find-me"` in `test.dts` as well. I could fix it up to do that later though, so that this can get in.

mike-scott commented:
> And yeah, that's piggybacking on an existing test. Could have a dedicated `test-bindings2/find-me-yaml` that matches with a new `compatible = "find-me"` in `test.dts` as well. I could fix it up to do that later though, so that this can get in.

@ulfalizer I think we want the test to run both ways so we're validating both paths: 1) first with 1 dir and then 2) with 2 dirs.  To do that we need a full set of matching bindings for each location.

ulfalizer commented:
@mike-scott 
There are ways to do it that doesn't involve creating any copies of large files.

If you back out the test code, I could push a small test to this branch. Is that okay?

mike-scott commented:
> @mike-scott
> There are ways to do it that doesn't involve creating any copies of large files.
> 
> If you back out the test code, I could push a small test to this branch. Is that okay?

@ulfalizer sounds good.  I'll drop the test changes (including the list fix).

ulfalizer committed:
scripts: dts: Add test for multiple binding directories

Add two bindings

    test-bindings/multidir.yaml
    test-bindings-2/multidir.yaml

and a new test-multidir.dts with two nodes that use them.

Verify that the two bindings were found by checking the
Device.binding_path attribute for the two device nodes.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

mike-scott commented:
Test removed.

ulfalizer commented:
@mike-scott 
Standalone test added.

ulfalizer reviewed (approved):
Code changes look okay to me. Can't review my own test though...

mike-scott commented:
@ulfalizer LGTM.  Nice and clean test for the 2nd directory.  Mine was rather messy.
